### ISSUE TITLE
mysql: DB notifications sometimes stored with escaped double quotes

### ISSUE THREAD

**Issue by OscarBarrett** [excid3/noticed#185]
In our deployment environments we've noticed that the notification `params` are stored with quotes escaped.
We haven't been able to work out why, as the DB schema is exactly the same, MySQL version is the same (5.7.33) and we're using docker in both dev + deployment environments to ensure the same ruby environment, OS etc.

This is a problem when using `has_noticed_notifications` as the `JSON_CONTAINS` condition doesn't match, so notifications aren't destroyed.

e.g.    
`"{\"num_vehicles\":1,\"funnel\":\"tender\",\"_aj_symbol_keys\":[\"num_vehicles\",\"funnel\"]}"`

Instead of    
`{"num_vehicles":1,"funnel":"tender","_aj_symbol_keys":["num_vehicles","funnel"]}`

A temporary workaround that we've employed is to wrap the arguments to `JSON_CONTAINS` with `JSON_UNQUOTE`.

i.e.
```ruby
model.where("JSON_CONTAINS(JSON_UNQUOTE(params), JSON_UNQUOTE(?))", Noticed::Coder.dump(param_name.to_sym => self).to_json)
```

**Comment by aitbw** [comment]
I was about to open an issue for the exact same problem; we also experienced this under our staging/production environments, but **only** for our first 43 (notification) records — beyond them, this hasn't occurred again, and we haven't been able to reproduce it under development either

We're using PG v11.13 and noticed 1.4.1

**Comment by excid3** [comment]
Able to write a test that replicates it?

**Comment by aitbw** [comment]
@excid3, hard to say; we've been monitoring our affected environments ever since this happened, but we weren't able to trace anything back to the root cause. I'm gonna attempt to write a test that replicates this behaviour though, but I can't promise anything :grimacing: maybe @OscarBarrett is able to do so?

I could provide an obfuscated version of the broken notifications, but doubt that's gonna be useful

Thanks in advance for the help, @excid3! And thanks for the great work you've been doing with noticed :tada:

**Comment by OscarBarrett** [comment]
I worked out our cause.

We were monkeypatching `table_exists?` so that noticed wouldn't block us during asset precompile when we don't have a DB available (looks like this is resolved since 1.5.2).    
`table_exists?` was returning false due to a baked in environment variable in the deployment images, so `Notification.noticed_coder` was falling back to `Noticed::TextCoder`.

So this can happen if `Noticed::TextCoder` ends up being used instead of `Noticed::Coder`.

Looking at the model generator, in older versions it could be possible to end up with a migration that created params as a text field instead of json which would also result in this issue.

Example failing test case: https://github.com/OscarBarrett/noticed/commit/0d437815d931482212761682fcdcad57f07c789e
Example fix for mysql that passes the above test: https://github.com/OscarBarrett/noticed/commit/21f80811f3149a5634482837948feef2ca441447

**Comment by excid3** [comment]
Good to know. Thanks for the update @OscarBarrett!
